### PR TITLE
WIP scheduler simulation

### PR DESCRIPTION
**Pull request by bkutil**
A simulation of a data store, adapter (`fetch` and `drain`) and a scheduler.

The idea is that the scheduler tries to emit all points older than the 'least recent start of a window for all adapters' and also that adapters have a cap on the data they can store in the buffer.

*NOT FOR MERGE*

### CODE REVIEW

**Comment by bkutil** [comment]
cc: @demmer @davidvgalbraith

File: extlib/scheduler-playground/lib/scheduler.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                + e.head.time.getTime()
+                + ', '
+                + e.tail.time.getTime()
+                + '), '
+                + name
+                + ' points: '
+                + e.size
+                + ' qlength: '
+                + e.qlength
+                );
```
Comment (VladVega) [inline extlib/scheduler-playground/lib/scheduler.js]: Consider trying out https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/template_strings here.

File: extlib/scheduler-playground/lib/store.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    this.data = [];
+
+    this.init();
+};
+
+Store.prototype.init = function() {
+    var range = this.last.getTime() - this.first.getTime();
+    var maxPoints = (range / 1000) * this.density;
+
+    while (this.data.length < maxPoints) {
```
Comment (VladVega) [inline extlib/scheduler-playground/lib/store.js]: Density is great parameter to be able to pass. Question: what happens when the maxPoints threshold is reached with all points representing of the first quarter of the time range requested?

File: extlib/scheduler-playground/lib/scheduler.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    var nextFetch = this.getNextFetch();
+
+    // All busy or finished
+    if (!nextFetch) { return; }
+
+    var cur = new Date(Date.now());
+
+    // If <to> is in the future, fetch until now
+    // This is where the window logic would go
+    if (nextFetch.to.getTime() > cur.getTime()) {
```
Comment (VladVega) [inline extlib/scheduler-playground/lib/scheduler.js]: Question about historical and super query:
Will there be a need to break the scheduled requests into windows for the historical parts of the query? Or will it always be `from` until `now > to ? to : now`? And then scheduled time windows for live.

File: extlib/scheduler-playground/lib/scheduler.js (hunk 1 above)
Comment (bkutil) [inline extlib/scheduler-playground/lib/scheduler.js]: This is cool, thanks! :)

File: extlib/scheduler-playground/lib/store.js (hunk 2 above)
Comment (bkutil) [inline extlib/scheduler-playground/lib/store.js]: What happens now is, that the number of points will always be `maxPoints` in the time interval - the average will equal to `density`, but the times are generated randomly (from a uniform distribution between `first` and `last`. That they would all fall into a sub-interval is possible, but won't happen on the average.

In that case, the adapter would just crunch the points in the first quarter as fast as it's queue size allows, then process the remaining few (or zero) points by querying with `from/to = (1/4, 1)`

Does that answer your question? I think that the comment above the `density` option is misleading and I should probably change that.

Also, for debugging purposes, to be able to specify a more regular sequence (e.g. tick every nth of a second) would be useful.

**Comment by dmajda** [comment]
Closing, the real implementation will be done in #167.

**Pull request closed without merging**